### PR TITLE
Autodetect input file level numbering, add consistency checks, and speed up transition parsing

### PR DESCRIPTION
**Pull request by lukeshingles**
## Level numbering autodetection

The starting level index of the atomic data input files (`groundstate_index_in`) was hardcoded to 1. It is now autodetected (0 or 1) from the first level of the first ion in `adata.txt` and logged. Since only the node-leader ranks parse `adata.txt` while all ranks parse level indices in the phixs and autoionisation files, the detected value is broadcast over the node communicator (and asserted to be 0 or 1 on every rank) before those files are read.

## Consistency checks against the detected numbering

Each check reports Z, ionstage, and the offending file before aborting:

- **`transitiondata.txt`**: a transition whose upper level lies beyond the ion's level count in `adata.txt` is a hard error (catches a 1-indexed transition file paired with 0-indexed `adata.txt`; the opposite shift is caught by the existing `lower >= 0` assertion). An ion whose transitions never have the ground state as lower level gets a warning, since sparse tables are legitimate (missing pairs are filled in).
- **phixs files**: every ion with any photoionisation data must include a ground state cross-section, so an excited-level table without one is a hard error. The check records the lowest supplied level per ion and per file *before* the retained-level filtering, so a misnumbered ground state table cannot be silently filtered out, and a correctly numbered companion file cannot mask a mismatched one. An ion whose ground state table comes from the companion phixs file is allowed with a warning.
- **`autoion.txt`**: the autoionising levels of an ion must form a contiguous block ending at its highest level, so the highest autoionising level index in the file must be the ion's top level — a shifted file aborts before any rates are stored.

## Faster transition table parsing

The per-row `istringstream` parsing (stringbuf copy plus locale-aware extraction) is replaced by whitespace tokenisation with `std::from_chars`, and the `std::set` used to detect existing transitions (a red-black-tree node allocation per row) is replaced by a binary search of the sorted table. Parsing a row drops from 400 ns to 219 ns (1.8x) in a microbenchmark over the kilonova dataset's 1.9 million transition rows.

The new parser preserves stream-extraction semantics for edge cases: leading `+` signs are accepted, underflow (even below double range, e.g. `1e-400`) is stored as zero, and overflow and `nan`/`inf` spellings are rejected. Covered by an 18-case edge test.

## Verification

Using the kilonova_1d regression setup (`REPRODUCIBLE=ON FASTMATH=OFF MAX_NODE_SIZE=2`, full newrun + resume + exspec sequence):

- Baseline vs. this branch on the standard 1-indexed dataset: all output files checksum-identical, so **no checksum regeneration is needed** (`from_chars` and stream extraction both produce correctly rounded floats).
- This branch on a copy of the dataset converted to 0-indexed (adata, transitiondata, and phixs level indices all shifted down by one): detected as 0-indexed and produced outputs checksum-identical to the 1-indexed run.
- Negative tests: mixing 0-indexed `adata.txt` with 1-indexed `transitiondata.txt` or 1-indexed `phixsdata_v2.txt` aborts during input reading with the specific error messages above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: input.cc [hunk 1]
```diff
@@ -1591,6 +1621,9 @@ void read_atomicdata_files() {
   }
   MPI_Barrier_node();
 
+  // only the node leaders read adata.txt, but all ranks parse level indices in autoion.txt and the phixs files
+  MPI_Bcast_safe(groundstate_index_in, 0, globals::mpi_comm_node);
+
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc:1746]: Added `assert_always(groundstate_index_in == 0 || groundstate_index_in == 1)` on all ranks immediately after the broadcast in 9d3c0b80. (Detection itself already asserts 0/1 before setting the value, so the only path to a bad broadcast was a run with zero included ions, which is unsupported anyway — the assert now fails fast in that case too.)

File: input.cc [hunk 2]
```diff
@@ -476,6 +484,16 @@ void read_ion_transitions(std::istream& ftransitiondata, const int ion_transitio
...
 
+  // if the ion has any transitions then the ground state must be the lower level of at least one of them,
+  // otherwise the level numbering in transitiondata.txt disagrees with the numbering detected from adata.txt
+  if (ion_transition_count_in_file > 0 && !found_groundstate_lower) {
+    printlnlog(
+        "[error] transitiondata.txt: no transition has the ground state (level index {} in the input files) as its "
+        "lower level",
+        groundstate_index_in);
+    assert_always(false);
+  }
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Done in 9d3c0b80 — `read_ion_transitions` now takes the atomic number and ionstage, and both the warning and the new out-of-range error name Z, ionstage, and the offending transition.

File: input.cc [hunk 3]
```diff
@@ -476,6 +484,16 @@ void read_ion_transitions(std::istream& ftransitiondata, const int ion_transitio
         {.lower = lower, .upper = upper, .A = A, .coll_str = coll_str, .forbidden = (intforbidden == 1)});
   }
 
+  // if the ion has any transitions then the ground state must be the lower level of at least one of them,
+  // otherwise the level numbering in transitiondata.txt disagrees with the numbering detected from adata.txt
+  if (ion_transition_count_in_file > 0 && !found_groundstate_lower) {
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc:553]: Good point — the fill-in of missing transitions below means ground-state participation isn't a hard invariant. In 9d3c0b80 this check is demoted to a warning, and a hard error was added on an actual index boundary instead: a transition whose upper level index is beyond the ion's level count in adata.txt (`upper >= nlevels_in_file`) aborts with a message naming Z, ionstage, and the transition. That catches a 1-based transition file paired with 0-based adata (verified with a deliberately mismatched dataset) without rejecting valid sparse tables; the opposite mismatch is still caught by the existing `lower >= 0` assertion.

File: input.cc [hunk 4]
```diff
@@ -1136,6 +1154,18 @@ void read_phixs_data() {
       // below is just an extra warning consistency check
       const int nlevels_groundterm = globals::elements[element].ions[ion].nlevels_groundterm;
 
+      // if any level of the ion has a photoionisation cross-section then the ground state must have one too,
+      // otherwise the phixs data numbers levels differently from adata.txt
+      if (get_nphixstargets(element, ion, 0) == 0 &&
+          std::ranges::any_of(std::views::iota(0, get_nlevels(element, ion)),
+                              [element, ion](const int level) { return get_nphixstargets(element, ion, level) > 0; })) {
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Agreed — `setup_phixs_list()` explicitly tolerates ions without a ground-level continuum, so this is demoted to a warning in 9d3c0b80. For an actually mismatched phixs file the warning identifies the likely cause, and the run still fails fast downstream (in testing, `ratecoeff.cc`'s `std::isfinite(alpha_sp)` assertion trips during rate coefficient precalculation), so nothing proceeds silently.

[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Correction to the above: an ion must have a ground state photoionisation cross-section unless it has no photoionisation data at all, so excited-level-only cross-sections for an ion are not a valid dataset and this condition is unambiguous evidence of mismatched numbering. The check is restored to a hard error in the latest commit.

File: input.cc [hunk 5]
```diff
@@ -424,19 +430,41 @@ void read_ion_levels(std::istream& adata, const int element, const int ion, cons
...
+  }
+  const auto tokenend = std::min(remainder.find_first_of(whitespace, tokenstart), remainder.size());
+  const auto token = remainder.substr(tokenstart, tokenend - tokenstart);
+#pragma clang unsafe_buffer_usage begin
+  const auto* const tokenfirst = token.data();
+  const auto* const tokenlast = token.data() + token.size();
+#pragma clang unsafe_buffer_usage end
+  const auto [ptr, ec] = std::from_chars(tokenfirst, tokenlast, value);
+  if (ec != std::errc{} || ptr != tokenlast) {
+    return false;
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc:485]: Fixed in 4cfa402b: a leading plus sign is now skipped before from_chars, and a token that parses but is out of float range is reparsed with double range and rounded to float, so underflowing values become zero as before. Covered by an edge-case test (+1.0e5, 1e-50, bare +, 1e999) and the kilonova_1d outputs remain bit-identical.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 6]
```diff
@@ -1136,6 +1201,19 @@ void read_phixs_data() {
       // below is just an extra warning consistency check
       const int nlevels_groundterm = globals::elements[element].ions[ion].nlevels_groundterm;
 
+      // every ion with any photoionisation data must include a cross-section for the ground state, so an excited
+      // level having a cross-section without the ground state means the phixs data numbers levels differently from
+      // adata.txt
+      if (get_nphixstargets(element, ion, 0) == 0 &&
+          std::ranges::any_of(std::views::iota(0, get_nlevels(element, ion)),
+                              [element, ion](const int level) { return get_nphixstargets(element, ion, level) > 0; })) {
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Fixed in cd8da41a: read_phixs_file now records per included ion whether the file supplies any table and a ground-index table, before the retained-level filtering, and read_phixs_data errors on any ion with tables but no ground state table. The previous post-filter any_of check is replaced by this raw-file check, so the single-ionising-level case can no longer slip through.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 7]
```diff
@@ -424,19 +432,55 @@ void read_ion_levels(std::istream& adata, const int element, const int ion, cons
...
+  const auto [ptr, ec] = std::from_chars(tokenfirst, tokenlast, value);
+  if (ec != std::errc{} || ptr != tokenlast) {
+    if constexpr (std::same_as<T, float>) {
+      // a valid number outside float range (such as an underflowing A value): reparse with double range and
+      // round to float, giving zero for underflow as stream extraction did
+      if (ec == std::errc::result_out_of_range && ptr == tokenlast) {
+        double dvalue{};
+        const auto [dptr, dec] = std::from_chars(tokenfirst, tokenlast, dvalue);
+        if (dec == std::errc{} && dptr == tokenlast) {
+          value = static_cast<float>(dvalue);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Fixed in cd8da41a: the double-range fallback now only accepts values with |value| <= FLT_MAX, so overflow like 1e100 is rejected (matching stream extraction) while underflow still rounds to zero. Covered by 1e100/-1e100 edge-case tests.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 8]
```diff
@@ -1079,16 +1155,40 @@ void read_phixs_data() {
...
-    read_phixs_file(2, tmpallphixs, tmpallphixstargets);
+    read_phixs_file(2, tmpallphixs, tmpallphixstargets, ion_has_phixstable_infile,
+                    ion_has_groundstate_phixstable_infile);
     MPI_Barrier_node();
   }
 
   if (phixs_file_version_exists[1]) {
-    read_phixs_file(1, tmpallphixs, tmpallphixstargets);
+    read_phixs_file(1, tmpallphixs, tmpallphixstargets, ion_has_phixstable_infile,
+                    ion_has_groundstate_phixstable_infile);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Fixed in a5eed89b: the tracking vectors and the ground-state check now live inside read_phixs_file, so each phixs file is checked against its own contents and a correctly numbered file can no longer mask a mismatch in the other. The error message names the offending file.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 9]
```diff
@@ -424,19 +443,55 @@ void read_ion_levels(std::istream& adata, const int element, const int ion, cons
...
+  const auto* const tokenlast = std::to_address(token.cend());
+  const auto [ptr, ec] = std::from_chars(tokenfirst, tokenlast, value);
+  if (ec != std::errc{} || ptr != tokenlast) {
+    if constexpr (std::same_as<T, float>) {
+      // a valid number below float range (such as an underflowing A value): reparse with double range and
+      // round to float, giving zero for underflow as stream extraction did. Overflow is still rejected
+      if (ec == std::errc::result_out_of_range && ptr == tokenlast) {
+        double dvalue{};
+        const auto [dptr, dec] = std::from_chars(tokenfirst, tokenlast, dvalue);
+        if (dec == std::errc{} && dptr == tokenlast && std::fabs(dvalue) <= std::numeric_limits<float>::max()) {
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Fixed in a5eed89b: the out-of-range fallback now uses strtod, which returns zero for underflow at any depth (including 1e-400, below double range) and HUGE_VAL for overflow, so underflow is stored as zero like stream extraction while overflow is still rejected via the |value| <= FLT_MAX guard. Covered by 1e-400 edge-case tests for float and double.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 10]
```diff
@@ -424,19 +443,55 @@ void read_ion_levels(std::istream& adata, const int element, const int ion, cons
...
+    return false;
+  }
+  const auto tokenend = std::min(remainder.find_first_of(whitespace, tokenstart), remainder.size());
+  auto token = remainder.substr(tokenstart, tokenend - tokenstart);
+  if (token.front() == '+') {  // stream extraction accepted a leading plus sign, but from_chars does not
+    token.remove_prefix(1);
+  }
+  const auto* const tokenfirst = token.data();
+  const auto* const tokenlast = std::to_address(token.cend());
+  const auto [ptr, ec] = std::from_chars(tokenfirst, tokenlast, value);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc:461]: Fixed in a5eed89b: every successful floating-point parse is now checked with |value| <= numeric_limits<T>::max(), which rejects nan (comparison false) and inf spellings before they can propagate. Covered by nan/inf/-inf edge-case tests.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 11]
```diff
@@ -384,6 +402,26 @@ void read_phixs_file(const int phixs_file_version, std::vector<float>& tmpallphi
...
 
+  // every ion with any photoionisation data must include a cross-section from the ground state, so an excited level
+  // having one without the ground state means this file numbers levels differently from adata.txt. Checked on the
+  // raw per-file contents so that a misnumbered table cannot be filtered out silently, and so that a correctly
+  // numbered second phixs file cannot mask a mismatch in this one
+  for (int element = 0; element < get_nelements(); element++) {
+    const int nions = get_nions(element);
+    for (int ion = 0; ion < nions; ion++) {
+      const auto uniqueionindex = get_uniqueionindex(element, ion);
+      if (ion_has_phixstable_infile[uniqueionindex] && !ion_has_groundstate_phixstable_infile[uniqueionindex]) {
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc]: Fixed in 1c2e2c74: an ion whose ground state table comes from the companion phixs file now gets a warning (naming both files) instead of an abort, so complementary v2+v1 datasets are accepted. The hard error remains when no file supplies the ground state table for an ion with any photoionisation data, and evidence stays per-file so the message names the offending file.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: input.cc [hunk 12]
```diff
@@ -1591,6 +1697,10 @@ void read_atomicdata_files() {
   }
   MPI_Barrier_node();
 
+  // only the node leaders read adata.txt, but all ranks parse level indices in autoion.txt and the phixs files
+  MPI_Bcast_safe(groundstate_index_in, 0, globals::mpi_comm_node);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.cc:1744]: Fixed in 1c2e2c74: read_autoion_data now records the highest raw autoionising level index per included ion and requires it to equal the ion's top level index in the file's numbering (the contiguous-to-top structure already asserted makes this a hard invariant). A shifted autoion.txt now aborts with an explicit numbering error before any rates are stored, instead of relying on the range/contiguity assertions.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
